icu4c/source/common/dictbe.cpp

```diff
@@ -1342,11 +1376,22 @@ CjkBreakEngine::divideUpDictionaryRange( UText *inText,
         t_boundary.addElement(numCodePts, status);
         numBreaks++;
     } else {
-        for (int32_t i = numCodePts; i > 0; i = prev.elementAti(i)) {
+        t_boundary.addElement(numCodePts, status);
+        numBreaks++;
+        int32_t prevIdx = numCodePts;
+
+        for (int32_t i = prev.elementAti(numCodePts); i > 0; i = prev.elementAti(i)) {
+            int length = prevIdx -  i;
+            prevIdx = i;
+            //skip the breaking point if it belongs to the particle or Hiragana.
+            if(isPhraseBreaking
+                    && (fParticleSet.containsKey(inString.tempSubString(i, length))
```

Comment:
As I mentioned in ICU4J review, could we merge fParticleSet and  fHiraganaWordSet into one fSkipSet and check only once by calling fSkipSet.containsKey() ? 
1. Remove fHiraganaWordSet from member data and the check here.
2. Rename fParticleSet to fSkipSet
3. Construct a on stack hiraganaWordSet in the constructor as how you  construct fHiraganaWordSet now but could be destrory after the constructor.
4. Call https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1UnicodeSetIterator.html w/ hiraganaWordSet and add the iteration to fSkipSet before or after loading the paractical from resources.